application/src/ext/java/org/opentripplanner/ext/carpooling/internal/CarpoolItineraryMapper.java

```diff
@@ -0,0 +1,153 @@
+   *
+   * <h3>Time Calculation Details</h3>
+   * <p>
+   * The method calculates three key times:
+   * <ol>
+   *   <li><strong>Driver pickup arrival:</strong> Driver's start time + pickup segment durations</li>
+   *   <li><strong>Passenger start:</strong> max(requested time, driver arrival time)</li>
+   *   <li><strong>Passenger end:</strong> start time + shared segment durations</li>
+   * </ol>
+   *
+   * <h3>Null Return Cases</h3>
+   * <p>
+   * Returns {@code null} if the candidate has no shared segments, which should never happen
+   * for valid insertion candidates but serves as a safety check.
+   *
+   * @param request the original routing request containing passenger preferences and timing
+   * @param candidate the insertion candidate containing route segments and trip details
+   * @return an itinerary with a single carpool leg, or null if shared segments are empty
+   *         (should not occur for valid candidates)
+   */
+  public Itinerary toItinerary(RouteRequest request, InsertionCandidate candidate) {
+    // Get shared route segments (passenger pickup to dropoff)
+    var sharedSegments = candidate.getSharedSegments();
+    if (sharedSegments.isEmpty()) {
+      return null;
+    }
+
+    // Calculate times
+    var pickupSegments = candidate.getPickupSegments();
+    Duration pickupDuration = Duration.ZERO;
+    for (var segment : pickupSegments) {
+      pickupDuration = pickupDuration.plus(
+        Duration.between(segment.states.getFirst().getTime(), segment.states.getLast().getTime())
+      );
+    }
+
+    var driverPickupTime = candidate.trip().startTime().plus(pickupDuration);
+
+    // Passenger start time is max of request time and when driver arrives
+    var startTime = request.dateTime().isAfter(driverPickupTime.toInstant())
+      ? request.dateTime().atZone(ZoneId.of("Europe/Oslo"))
+      : driverPickupTime;
+
+    // Calculate shared journey duration
+    Duration carpoolDuration = Duration.ZERO;
+    for (var segment : sharedSegments) {
+      carpoolDuration = carpoolDuration.plus(
+        Duration.between(segment.states.getFirst().getTime(), segment.states.getLast().getTime())
+      );
+    }
+
+    var endTime = startTime.plus(carpoolDuration);
+
+    // Get vertices from first and last segment
+    var firstSegment = sharedSegments.get(0);
+    var lastSegment = sharedSegments.get(sharedSegments.size() - 1);
+
+    Vertex fromVertex = firstSegment.states.getFirst().getVertex();
+    Vertex toVertex = lastSegment.states.getLast().getVertex();
+
+    // Collect all edges from shared segments
+    var allEdges = sharedSegments.stream().flatMap(seg -> seg.edges.stream()).toList();
+
+    // Create carpool leg
+    CarpoolLeg carpoolLeg = CarpoolLeg.of()
+      .withStartTime(startTime)
+      .withEndTime(endTime)
+      .withFrom(Place.normal(fromVertex, new NonLocalizedString("Carpool boarding")))
+      .withTo(Place.normal(toVertex, new NonLocalizedString("Carpool alighting")))
```

Comment:
You could translate these with internals.properties thingy, but that's up to you.